stan/math/prim/mat/fun/qr_thin_Q.hpp

```diff
@@ -0,0 +1,30 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_QR_THIN_Q_HPP
+#define STAN_MATH_PRIM_MAT_FUN_QR_THIN_Q_HPP
+
+#include <stan/math/prim/arr/err/check_nonzero_size.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/scal/err/check_greater_or_equal.hpp>
+#include <Eigen/QR>
+#include <algorithm>
+
+namespace stan {
+namespace math {
+
+template <typename T>
+Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> qr_thin_Q(
+    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& m) {
+  typedef Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> matrix_t;
+  check_nonzero_size("qr_Q", "m", m);
+  Eigen::HouseholderQR<matrix_t> qr(m.rows(), m.cols());
+  qr.compute(m);
+  const int min_size = std::min(m.rows(), m.cols());
+  matrix_t Q = qr.householderQ() * matrix_t::Identity(m.rows(), min_size);
```

Comment:
Copy instead of multiplying by an identity.

Reply:
Unless ofc. that's what Eigen does this already with template magic... Tricky Eigen :/

Reply:
This is how the Eigen doc says to do it. I don't think the `qr.householderQ()` matrix expression is copy-able or subset-able. But multiplying something by `matrix_t::Identity()` is overloaded so that it does not actually multiply anything.